File: src/common/ts/src/message-broker/factories/producer-factory.ts
import { IProducer } from '../interfaces/iproducer';
import { IProducerConfiguration } from '../interfaces/iproducer-configuration';
import { KafkaProducer } from '../kafka/kafka-producer';

export const ProducerType = {
  KAFKA: 'Kafka',
};

// Helps with creation of a IProducer based on
// the configuration sent
export class ProducerFactory {
  private static kafkaProducers: Record<string, IProducer> = {};

  static getProducer(metadata: IProducerConfiguration): IProducer | undefined {
    if (metadata.type === ProducerType.KAFKA) {
      return this.getKafkaProducer(metadata.broker);
    }
    return undefined;
  }

  private static getKafkaProducer(broker: string): IProducer {
    if (this.kafkaProducers[broker] !== undefined) {
      return this.kafkaProducers[broker];
    } else {
      const producer = KafkaProducer.getInstance(broker);
      this.kafkaProducers[broker] = producer;
      return producer;
    }
  }
}

File: src/common/py/casho/casho/kafka/factories/producer_factory.py
from ..interfaces import IProducer, IProducerConfiguration
from ..Kafka_implementation.kafka_producer import KafkaProducer

class ProducerFactory:
    kafka_producers = {}

    @staticmethod
    def get_producer(metadata: IProducerConfiguration) -> IProducer or None:
        if metadata.type == "Kafka":
            return ProducerFactory.get_kafka_producer(metadata.broker)
        return None

    @staticmethod
    def get_kafka_producer(broker: str) -> IProducer:
        if broker in ProducerFactory.kafka_producers:
            return ProducerFactory.kafka_producers[broker]
        else:
            producer = KafkaProducer.get_instance(broker)
            ProducerFactory.kafka_producers[broker] = producer
            return producer
